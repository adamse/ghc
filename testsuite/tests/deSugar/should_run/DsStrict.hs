{-# LANGUAGE Strict #-}
{-# OPTIONS_GHC -fno-warn-unused-binds #-}
module Main where
import Debug.Trace



f0 a = "fun"
f0' ~a = "fun'"

f1 ~n = case n of
          a -> "case"
f1' ~n = case n of
           ~a -> "f1'"


f2 = \a -> "lamda"
f2' = \ ~a -> "lambda'"

f3 ~n = let a = n
        in "let"
f3' ~n = let ~a = n
         in "let'"


newtype Age = MkAge Int

f4, f4' :: Age -> String
f4 (MkAge a) = "newtype"
f4' ~ (MkAge a) = "newtype'"

main :: IO ()
main = mapM_ (\(n, f) -> putStrLn (f (v n))) fs
  where fs = [("fun", f0)
             ,("fun'", f0')
             ,("case", f1)
             ,("case'", f1')
             ,("lambda", f2)
             ,("lambda'", f2')
             ,("let", f3)
             ,("let'", f3')
             ,("newtype", (\ ~i -> f4 (MkAge i)))
             ,("newtype'", (\ ~i -> f4' (MkAge i)))]
        v n = trace ("evaluated in " ++ n) 1
