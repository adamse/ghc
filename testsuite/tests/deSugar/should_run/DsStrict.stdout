fun
fun'
case
f1'
lamda
lambda'
let
let'
newtype
newtype'
